marlowe-dashboard-client/src/Contract/Types.purs

```diff
@@ -14,31 +14,41 @@ import Data.Map (Map)
 import Data.Maybe (Maybe(..))
 import Data.Set (Set)
 import Halogen (RefLabel(..))
-import Marlowe.Execution (ExecutionState, NamedAction)
+import Marlowe.Execution.Types (NamedAction)
+import Marlowe.Execution.Types (State) as Execution
 import Marlowe.Extended.Metadata (MetaData)
 import Marlowe.PAB (PlutusAppId, MarloweParams)
 import Marlowe.Semantics (ChoiceId, ChosenNum, Party, Slot, TransactionInput, Accounts)
 import WalletData.Types (WalletDetails, WalletNickname)
 
+-- The Contract.State wraps the Execution.State up with some additional information that is necessary
+-- to render the execution state of the contract in the app. Some of this is derived data that could
+-- be calculated on the fly, but is stored here for efficiency. Some of it is data specific to the UX
+-- (like which step card is currently centered, and which tabs are shown in which step cards). And some
+-- of it is metadata (the actual metadata of the contract, but also functional metadata like the ID of
+-- the user's follower app for this contract, and the MarloweParams of the contract).
 type State
-  = { tab :: Tab -- this is the tab of the current (latest) step - previous steps have their own tabs
-    , executionState :: ExecutionState
-    , previousSteps :: Array PreviousStep
-    , followerAppId :: PlutusAppId
+  = { followerAppId :: PlutusAppId
     , marloweParams :: MarloweParams
-    -- Which step is selected. This index is 0 based and should be between [0, previousSteps.length]
-    -- (both sides inclusive). This is because the array represent the past steps and the
-    -- executionState has the current state and visually we can select any one of them.
-    , selectedStep :: Int
     , metadata :: MetaData
+    , executionState :: Execution.State
+    -- This is the tab of the current (latest) step card - previous step cards have their own tabs.
+    , tab :: Tab
+    -- These contain the data needed to render the previous step cards. This corresponds loosely to the
+    -- array of PreviousStates in the Execution.State, except that the shape and content of the data is
+    -- slightly different, and we include timeout steps here in addition to transaction steps.
```

Comment:
In general, I'd try to avoid explaining the `what` if it can be "easily" understood and focus on the `why` if it's not clear.

Most of this comment focus on the `what`, which can be seen right below this data type. On the `why` we need to maintain a similar structure to the Execution.State, I'm sorry, I should have explicit it when I introduced the change.

At some point we only used to have the ExecutionState to render the past and current card, but when I implemented the timeout card, I realized that we needed to decouple the cards from the execution state. The reason is that if a card has timeouted, the actual contract does not move, it keeps its current state until you make a transaction, which can be an empty transaction just to "advance", or if the timeout has a `When` contract it can be one of the actions in the `Case` array.

We use `regenerateStepCards` to construct this array from the ExecutionState, which may include a `PendingTimeouts`. That object has a comment that kind of explain this, so we can refer to it.